### DISCUSSION TITLE
Inertia SSR with Laradock

### DISCUSSION THREAD

**Discussion by notAPanda** [inertiajs/inertia#1499] [Help]
Hey Everyone!

I wanted to use Inertia with Laradock.
I am unable to to make SSR to work. 
It works perfectly with sail.

I am running caddy, mysql, redis, php-worker containers.

I have tried to run artisan inertia:start-ssr inside the workspace container but the SSR is not rendering any html.
I know that SSR works on 13714 port. Do i need to expose this port in one of the containers?

Please let me know if anyone knows what to do. 
Cheers!

**Comment by dualklip** [comment]
I have the same problem. In my case I can say php-worker has not installed node. 

My config to run `php artisan inertia:start-ssr` with php-worker is:

```
[program:laravel-inertia-ssr]
process_name=%(program_name)s_%(process_num)02d
command=php /var/www/artisan inertia:start-ssr
autostart=true
autorestart=true
user=laradock
redirect_stderr=true
stdout_logfile=/var/www/storage/logs/inertia-ssr.log
```

As you can see I set the log file in `/var/www/storage/logs/inertia-ssr.log`, and this is what I receive on that file:

```
sh: exec: line 0: node: not found
sh: exec: line 0: node: not found
sh: exec: line 0: node: not found
```

The laravel log file says this:

```
production.ERROR: sh: exec: line 0: node: not found
 {"exception":"[object] (Inertia\\Ssr\\SsrException(code: 0): sh: exec: line 0: node: not found
 at /var/www/vendor/inertiajs/inertia-laravel/src/Commands/StartSsr.php:74)
[stacktrace]
#0 /var/www/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Inertia\\Commands\\StartSsr->handle()
#1 /var/www/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#2 /var/www/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure(Object(Closure))
#3 /var/www/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod(Object(Illuminate\\Foundation\\Application), Array, Object(Closure))
#4 /var/www/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\\Container\\BoundMethod::call(Object(Illuminate\\Foundation\\Application), Array, Array, NULL)
#5 /var/www/vendor/laravel/framework/src/Illuminate/Console/Command.php(194): Illuminate\\Container\\Container->call(Array)
#6 /var/www/vendor/symfony/console/Command/Command.php(312): Illuminate\\Console\\Command->execute(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Illuminate\\Console\\OutputStyle))
#7 /var/www/vendor/laravel/framework/src/Illuminate/Console/Command.php(164): Symfony\\Component\\Console\\Command\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Illuminate\\Console\\OutputStyle))
#8 /var/www/vendor/symfony/console/Application.php(1040): Illuminate\\Console\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#9 /var/www/vendor/symfony/console/Application.php(314): Symfony\\Component\\Console\\Application->doRunCommand(Object(Inertia\\Commands\\StartSsr), Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#10 /var/www/vendor/symfony/console/Application.php(168): Symfony\\Component\\Console\\Application->doRun(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#11 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(200): Symfony\\Component\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#12 /var/www/artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#13 {main}
"}
```

When I go inside of workspace `docker-compose exec --user=laradock workspace bash` and run the command manually works fine, but this is because workspace has node installed.

Somebody has modified php-worker Dockerfile to have node on this container?

**Reply by dualklip** [reply]
I'm testing the following change in `/php-worker/Dockerfile`

In line 22 change from this:
```
RUN apk --update add wget \
  curl \
  git \
  build-base \
  libmcrypt-dev \
  libxml2-dev \
  linux-headers \
  pcre-dev \
  zlib-dev \
  autoconf \
  cyrus-sasl-dev \
  libgsasl-dev \
  oniguruma-dev \
  supervisor; \
  if [ $(php -r "echo PHP_VERSION_ID - PHP_RELEASE_VERSION;") = "50600" ] || \
     [ $(php -r "echo PHP_VERSION_ID - PHP_RELEASE_VERSION;") = "70000" ]; then \
    apk --update add libressl libressl-dev; \
  else \
    apk --update add openssl-dev; \
  fi
```
to this
```
RUN apk --update add wget \
  curl \
  git \
  npm \
  build-base \
  libmcrypt-dev \
  libxml2-dev \
  linux-headers \
  pcre-dev \
  zlib-dev \
  autoconf \
  cyrus-sasl-dev \
  libgsasl-dev \
  oniguruma-dev \
  supervisor; \
  if [ $(php -r "echo PHP_VERSION_ID - PHP_RELEASE_VERSION;") = "50600" ] || \
     [ $(php -r "echo PHP_VERSION_ID - PHP_RELEASE_VERSION;") = "70000" ]; then \
    apk --update add libressl libressl-dev; \
  else \
    apk --update add openssl-dev; \
  fi
```

Added `npm` to the list. So far is working but is not well tested

**Comment by vsjoshtait** [comment]
Do I need to port forward the 13714 port? I have a php worker pod that runs `php artisan inertia:start-ssr` but no idea how to access the URL

**Reply by sonbarros** [reply]
you need to create config/inertia.php and point the address that is working in docker

**Reply by reinink** [reply]
Also, keep in mind that the SSR server isn't going to be public facing — as in you're unlikely going to be able to access it at a browser, but rather you're PHP service will connect locally to this HTTP service:

https://github.com/inertiajs/inertia-laravel/blob/master/src/Ssr/HttpGateway.php#L19

If you run `php artisan inertia:start-ssr` you should be able to http://localhost:13714/render to load locally (although it will generate an error in the SSR server because a JSON payload isn't provided).

**Reply by sonbarros** [reply]
Thank you reinink. I'm getting good results with inertiajs. gratitude

**Comment by DlouhyDavid** [comment]
I keep a strict separation of processes and node runs in a different container. Not where PHP is. Can I somehow set it to use a different host?

**Reply by SanderSander** [reply]
I had the same issue and had to change the `config/inertia/php` configuration. 

```

return [
    'ssr' => [
        'enabled' => true,
        'url' => env('SSR_URL', 'http://127.0.0.1:13714'),
        // 'bundle' => base_path('bootstrap/ssr/ssr.js'),
    ],
    
    ```

**Comment by abdessamadbettal** [comment]
check this repo https://github.com/abdessamadbettal/laravel-starter

Readme has all instructions how to install it and how to use

**Comment by cvairlis** [comment]
Hello all, I found the solution at the inertiajs discord community after a search for laradock.
Set the `inertia.ssr.url` at the `config/inertia.php` config as: `http://workspace:13714` and it may work.
After this change and running the ssr server you may see the rendered html at view page source.
Shout out to @krasoff from discord server of inertiajs.

**Reply by richt222** [reply]
Thanks @cvairlis . Note for others, this may also be true of other docker setups. A bit like setting the database host to `database` instead of localhost. It's internal (PHP) to internal (SSR server). So we're using an internal reference. In Laradock it is commonly `http://appserver:13714`
